v3/src/transports/http/__tests__/utils.test.ts

```diff
@@ -0,0 +1,310 @@
+   *
+   * This will not yield an actual handler until data has been transmitted
+   * to be used by the handler with `initHandlerWithParams` for the reasons
+   * noted in that method's description.
+   *
+   * In practice, since the handler doesn't actually return anything, this will
+   * either result in a resolution to `undefined`, or a `rejection` with an
+   * error.
+   */
+  handlerPromise: ReturnType<AsyncRequestListener>;
+}
+
+/**
+ * Generate a mock `RequestListener` using the [`node-mocks-http`] npm module
+ * in order to simulate how an `http.IncomingRequest` and `http.ServerResponse`
+ * would behave, including an `EventEmitter` for the response.
+ *
+ * [`node-mocks-http`]: https://npm.im/node-mocks-http
+ */
+function createMockRequestListener(
+  requestOptions: RequestOptions,
+  responseOptions: ResponseOptions = Object.create(null),
+) {
+  return {
+    req: createRequest({
+      ...requestOptions,
+    }),
+    res: createResponse({
+      eventEmitter: EventEmitter,
+      ...responseOptions,
+    })
+  }
+}
+
+/**
+ * Create a testable `RequestListener` to test both POST and GET requests
+ *
+ * POST requests receive their parameters as a steram of data on the body
+ * while GET requests are provided their parameters in the query string.  We
+ * aim to support both of these methods though the behavior between them should
+ * be mostly the same.  While POST requests might eventually be
+ * used for more complicated purposes (e.g. uploads), the core functionality
+ * (e.g. parsing of `variables`, `extensions`) should remain the same and be
+ * tested in an identical fashion.  The actual parsing of the parameters is
+ * still tested separately, within the `parseGetRequest` and `parsePostRequest`
+ * methods laster in this test-suite.
+ *
+ * @param params Parameters defined by `TestableRequestListenerParams`
+ */
+export function getTestableRequestListener(
+  {
+    requestOptions,
+    responseOptions,
+    handler,
+  }: {
+    requestOptions: RequestOptions;
+    responseOptions?: ResponseOptions;
+    handler: AsyncRequestListener;
+  }): TestableRequestListener {
+
+  if (!requestOptions.method) {
+    throw new Error(
+      "Internal error: Must pass `method` to " +
+      "`getTestableRequestListener`'s `requestOptions`.");
+  }
+
+  // Create a `Promise` which will eventually resolve to the return result of
+  // the aynchronous handler.  In order to send the data only after the handler
+  // was created, the handler won't actually be created until after we have
+  // data to be sent to it.
+  let handlerPromiseResolve: (value?: Promise<void>) => void;
+  const handlerPromise =
+    new Promise<ReturnType<AsyncRequestListener>>((resolve) => {
+      handlerPromiseResolve = resolve;
+    })
+    // We'll actually return the `Promise` of the handler itself here to
+    // un-wrap what would otherwise be a `Promise` of a `Promise`.
+    .then((resolved) => resolved);
```

Comment:
This throws me for a bit of a loop, but I can't work around it. I'd love to understand why this is necessary, but more importantly, why the following doesn't yield the same result (my understanding is off somewhere).

```ts
  const handlerPromise =
    Promise.resolve(
      new Promise<ReturnType<AsyncRequestListener>>((resolve) => {
        handlerPromiseResolve = resolve;
      })
    );
```